### DISCUSSION TITLE
EMQX over Kubernetes using HPA

### DISCUSSION THREAD

**Discussion by bloodberrys** [emqx/emqx#4715] [General]
Thanks for @zmstone, previously I sent him this question via email, and perhaps this topic would approach more benefit when we are using EMQX over Kubernetes. 

I just wondering and hopefully, I could understand more or validating my emqx understanding from the expert.
(sorry if any grammar mistake, since I'm not a native English speaker)

I had provisioned emqx using helm charts, and when it deployed, it will consist: 

a Statefulset with 3 pods by default, and 2 services (clusterIP and headless). 

After I patched the clusterIP service to become a LoadBalancer Type (whatever it is: internal or external, in my case I'm using Network Load Balancer on EKS which can handle million of TCP traffic) and If I decided to use HPA (horizontal pod autoscaler) on statefulsets to scale my emqx cluster when a lot of traffic comes to LoadBalancer (external/internal type) so it would become scalable, and set the minimum number of cluster with 1 number (or 1 pod) on HPA and hope when traffic goes low it only shows 1 pod, is it okay with that?

If okay, why? but what will happen? is there any trouble related to data logs or something?
If not okay, why? so we have to set the minimum with 3 (based on default), isn't it?

---
This is section 2:

Technically speaking, I had tried it with a publisher stress test (using MQTT JMETER with ~20k concurrent traffics) with only publisher event (so we only have "publish" event pretending an IoT device send their message to broker) and random cliend_id. It worked, traffic was distributed to all over cluster, even if I turn off the emqx bridge, but I saw a strange behavior in the dashboard. When I only emulate "publish" event with a lot of traffic, and HPA increase more pods, a brand-new pod didn't have any connection number (zero connection), so only the old pods will have connection number. A brand-new pod will have a connection number if and only if I stop the JMETER and restart the stress test... Is it normal behavior?

I really appreciate your work, Thank you.

**Comment by zmstone** [comment]
### HPA (horizontal pod autoscaler) on statefulsets to scale my emqx cluster

A disclaimer first: I have very limited experience with k8s, not to mention HPA.
The first question hits me is how do automate new pod join the cluster when scaling up,
and old pod leaving the cluster when scaling down.
I suppose you have made it happen, it would be great if you can share your experiences
 in another GitHub discussion under the category of "Ideas".

> If okay, why? but what will happen? is there any trouble related to data logs or something?

- We do not recommend over-sized cluster, most of the common use cases, 3 nodes is enough
  When scale up to many nodes, the network latency between the nodes becomes performance killer,
  and the database cluster has higher risk of brain-split
- Scale down to single node is OK. After all a cluster is bootstrapped from a single node.
- It should (auto) scale up before RAM usage is too high
  (i.e. do not only rely on CPU utilisation to make the decision).
  Because a newly joined node need to fetch all the data from one of the running nodes,
  this fetch may cause some overheads for the running node.
- Data loss risk is low, since the database is fully replicated to all nodes in the cluster
- Logs are not replicated. And starting from 4.3, EMQ X nodes running in containers
  will only log to console by default. Assuming people'd have log forwarder picking up
  the container logs for indexing

### Answer to "This is section 2:"

> When I only emulate "publish" event with a lot of traffic, and HPA increase more pods, a brand-new pod didn't have any connection number (zero connection), so only the old pods will have connection number. A brand-new pod will have a connection number if and only if I stop the JMETER and restart the stress test... Is it normal behavior?

EMQX nodes do not auto rebalance the load. There is no such feature yet.
i.e. EMQX nodes do not close tcp connections in hoping
that the clients will try to reconnect and reach another node.

**Reply by bloodberrys** [reply]
Thanks for the reply! @zmstone 

> I suppose you have made it happen, it would be great if you can share your experiences in another GitHub discussion under the category of "Ideas".

I'm very happy to say this but yes I did it and I will share this idea. Anyway, I'm going to run my experiment for scalable EMQX tomorrow or the day after it. Since I have +-500 vCPU in my current workload for my research, I will shot EMQX Broker by using distributed tests (Distributed MQTT JMETER), and hopefully, I can reach  **1 million connections**, I wish I will have a piece of good news, so wish me luck. Only for AWS - EKS use case. I'm still don't have any experience in using GCP - GKE and Azure - AKS but I'm pretty sure the rest of them have pretty much the same mechanism since Kubernetes is actually an independent tech concept. I will perhaps share "How to use EMQX in Kubernetes (EKS)" and the result as well after that on the discussion. I wish I have a "**tuning**" recommendation from you for such a big load and what features that I have to look at caused by a lot of connection on this experiment before I run it.

> We do not recommend over-sized cluster, most of the common use cases, 3 nodes is enough
**When scale up to many nodes, the network latency between the nodes becomes performance killer,
and the database cluster has higher risk of brain-split**

Interesting, I will consider this answer, but then again I will test it with ~1 million on Kubernetes.

> Scale down to single node is OK. After all a cluster is bootstrapped from a single node.

I see, so everything is gonna be merged, am I correct?

> It **should (auto) scale up before RAM usage** is too high
(i.e. **do not only rely on CPU utilisation** to make the decision).

This is really Insightful, in my current HPA, I'm still considering the _CPU_ for thresholding an auto-scaler, thank you.

> Data loss risk is low, since the database is fully replicated to all nodes in the cluster

What kind of DB emqx using to showing a number of data on the dashboard? Redis? memcached? I have to know it before the test.


> EMQX nodes do not auto rebalance the load. There is no such feature yet. i.e. EMQX nodes do not close TCP connections in hoping that the clients will try to reconnect and reach another node.

If EMQX nodes do not auto rebalance, is emqx bridge will do it? and will reach another nodes? I read it from here https://docs.emqx.io/en/broker/v4.2/advanced/bridge.html#mqtt-bridging-example it will forward a message from any broker to another broker, can we have a recommendation to tune this feature so it will work on this research use-case?

**Reply by zmstone** [reply]
> I'm very happy to say this but yes I did it and I will share this idea.

Thank you! we are looking forward to your sharing

> Since I have +-500 vCPU in my current workload for my research, I will shot EMQX Broker by using distributed tests (Distributed MQTT JMETER), and hopefully, I can reach 1 million connections, I wish I will have a piece of good news, so wish me luck. 

Great! We have done various tests with Enterprise edition, but I'm not sure if we have ever conducted such tests against k8s. Will check with my team mates.

> I wish I have a "tuning" recommendation from you for such a big load and what features that I have to look at caused by a lot of connection on this experiment before I run it.

We hope our default values should be generic enough for small and for large scale.
If you plan to test wildcard topics, we'd recommend you to try v4.3-rc.5
to which we have merged a few performance improvement PRs.

> What kind of DB emqx using to showing a number of data on the dashboard? Redis? memcached? I have to know it before the test.

EMQ X runs on Erlang's built-in DB named Mnesia.
Mnesia is mostly used to store routing information and some configurations.
Client information on dashboard is aggregated.
Most of the counters and gauges are collected from individual cluster members.

> If EMQX nodes do not auto rebalance, is emqx bridge will do it? and will reach another nodes? I read it from here https://docs.emqx.io/en/broker/v4.2/advanced/bridge.html#mqtt-bridging-example it will forward a message from any broker to another broker, can we have a recommendation to tune this feature so it will work on this research use-case?

If your goal is to reach 1M connection, I am pretty sure you do not need bridging.
Bridging is a good choice to connect two separate borkers or clusters.
The client information is not shared between the nodes,
meaning there is no way to make sure a client can not connect to two bridged brokers at the same time.
quite a lot other limitations too.

**Comment by zmstone** [comment]
BTW. We are working on a new feature called Rlog
The short summary of it is to allow EMQ X nodes to run in two different roles.
'Core' nodes and 'replicant' nodes.
Core nodes will run as a static cluster (typically no more than 3 nodes),
and replicant nodes are stateless, they should fit perfectly in k8s auto scaling deployments.

**Reply by k32** [reply]
Yes, it would be interesting to see how RLOG can autoscale.

**Reply by bloodberrys** [reply]
master-slave? sounds more intuitive. Can't wait for it. I wish I could test it on my workload 😂

**Reply by zmstone** [reply]
no, you can maybe call the core cluster master nodes, (they are for high availability)
then the replicant nodes, they randomly pick one core node to 'follow', for getting data replicated async-ly

**Reply by zmstone** [reply]
> master-slave? sounds more intuitive. Can't wait for it. I wish I could test it on my workload 😂

sure, we'll share news as we move forward, you are always welcome to test it out.

**Reply by k32** [reply]
P.S. Workbranch is here: https://github.com/k32/emqx/tree/rlog-ekka It should already support MQTT routing. If we can somehow collaborate on load testing with autoscaling, that would be great.

**Reply by bloodberrys** [reply]
> P.S. Workbranch is here: https://github.com/k32/emqx/tree/rlog-ekka It should already support MQTT routing.

Cool! noted and watched. Anyway, is there any architecture diagram available on this brand-new feature called RLOG? The new feature always has a big opportunity to grow. I would really appreciate including it on my research paper and hopefully published on the particular publisher, and referencing at EMQX pages/working repo and the collaborator/author's name for attribution.

**Reply by k32** [reply]
Sure, there are some docs here: https://github.com/emqx/eip/blob/main/active/0004-async-mnesia-change-log-replication.md

**Reply by n8o** [reply]
What is the current progress on adding this new feature?

**Reply by k32** [reply]
@n8o It has been merged to the unstable 5.0 branch. Please note that 5.0 is still in alpha stage.
